### ISSUE TITLE
Use setdefault in handleHeader

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4216]
|<img alt="johnnyg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| johnnyg reported|
|-|-|
|Trac ID|trac#4216|
|Type|enhancement|
|Created|2010-01-14 10:46:41Z|
|Branch|https://github.com/twisted/twisted/tree/handleheader-setdefault-4216|

In twisted/web/client.py, the function handleHeader (on L66) does:

```python
l = self.headers[key] = self.headers.get(key, [])
```

which is less clear and less efficient than doing:

```python
l = self.headers.setdefault(key, [])
```

I've also attached a patch if you're feeling lazy.

Attachments:

* [client.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3a5/3a5fc83f31558e73d2df0b537e4a628ec2c5aabb/869b338d49249cd742618fbcacc2a9fd9ab7e120.patch) (117 bytes) - added by johnnyg on 2010-01-14 10:47:07Z - 
* [4216_handleheader.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3a5/3a5fc83f31558e73d2df0b537e4a628ec2c5aabb/736e7273a67f0cce53f19673f1f750ac5bd55bff.patch) (684 bytes) - added by khorn on 2010-01-14 17:19:09Z - Updated patch with updated copyright header
* [4216_handleheader_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3a5/3a5fc83f31558e73d2df0b537e4a628ec2c5aabb/1cd9596b28ae2af93e4a9aa44350b4c1fac8d912.patch) (1279 bytes) - added by khorn on 2010-02-02 19:53:07Z - Updated patch to address review comments.

<details><summary>Searchable metadata</summary>

```
trac-id__4216 4216
type__enhancement enhancement
reporter__johnnyg johnnyg
priority__normal normal
milestone__ 
branch__branches_handleheader_setdefault_4216 branches/handleheader-setdefault-4216
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1263466001000000 1263466001000000
changetime__1265654331000000 1265654331000000
version__None None
owner__ 
cc__thijs cc__jesstess cc__khorn
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_2"></a>|@khorn commented|
|-|-|

My system didn't like johnnyg's patch, so I went ahead and created a new one, including updating the copyright header.

Hopefully this helps out the reviewer (I'd do it, but I somehow broke trial on my machine...and haven't managed to fix it yet)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_4"></a>|@jesstess set owner to @khorn|
|-|-|

See the comment from glyph in [#2062](https://github.com/twisted/twisted/issues/2062) about how twisted.web.client.Agent is hopefully the future and twisted.web.client is on the way out.

That said, treating this like a normal review:
 * handleHeader needs a docstring
 * the HTTPPageGetter class needs a docstring

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_5"></a>|@khorn removed owner|
|-|-|

Updated patch addresses review comments.

Sending to review.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess commented|
|-|-|

(In [28120]) Branching to 'handleheader-setdefault-4216'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess commented|
|-|-|

(In [28121]) Apply khorn's patch 4216_handleheader_2.patch, add @type and @param
markup in handleHeader, and add a topfile.

refs [#4216](https://github.com/twisted/twisted/issues/4216)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_9"></a>|@jesstess set owner to @khorn|
|-|-|

khorn, if you don't mind the markup I added I'll commit this to trunk (or you can since you have bits).

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_10"></a>|@khorn set owner to @jesstess|
|-|-|

jesstess: could you go ahead and commit it?  I'm trying to debug some Windows problems with Combinator and Combinator is in a very broken state at the moment on my machine.

BTW, the markup looks fine.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_11"></a>|@jesstess set status to `closed`|
|-|-|

(In [28130]) Merge handleheader-setdefault-4216

Authors: johnnyg, khorn, jesstess
Reviewers: jesstess, khorn
Fixes: [#4216](https://github.com/twisted/twisted/issues/4216)

Use setdefault instead of less clear code that does the same thing in
HTTPPageGetter.handleHeader.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
